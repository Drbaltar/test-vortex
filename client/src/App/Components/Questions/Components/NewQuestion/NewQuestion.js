import React from 'react';
import Axios from 'axios';

import QuestionForm from '../QuestionForm/QuestionForm';

class NewQuestion extends React.Component {
    constructor(props) {
        super(props);

        this.initialState = {
            submissionResponse: '',
            successAlert: false
        }

        this.state = this.initialState;
    }

    submitNewQuestion = (questionData) => {
        Axios.post('/api/questions/submit', questionData)
            .then((response) => this.setState({submissionResponse: response, successAlert: true}))
            .catch((response) => this.setState({submissionResponse: response}))
    }

    render() {
        let newQuestionView;

        if (this.state.successAlert) {
            newQuestionView = (
                <div className="alert alert-success" role="alert">
                    {this.state.submissionResponse.data}
                </div>
            )
        } else {
            newQuestionView = (
                <QuestionForm title="Add New Question (Patriot)"
                    submitEvent={(questionData) => this.submitNewQuestion(questionData)}/>
            )
        }
        return(
            <div>
                {newQuestionView}
            </div>
        );
    }
}

export default NewQuestion;