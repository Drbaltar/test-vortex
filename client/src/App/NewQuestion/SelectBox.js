import React from 'react';

const SelectBox = (props) => {
    let options = props.options;
    const selectableOptions =
        options.map(option => <option key={option}>{option}</option>)

    return(
        <div className="form-group">
            <label for={props.id}>{props.label}</label>
            <select className="form-control" id={props.id} onChange={props.inputChange}>
                {selectableOptions}
            </select>
        </div>
    )
}

export default SelectBox;