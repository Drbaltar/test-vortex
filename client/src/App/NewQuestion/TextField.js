import React from 'react';

const TextField = (props) => {
    return(
        <div className="form-group">
            <label for={props.id}>{props.label}</label>
            <input className="form-control" type={props.type} id={props.id} value={props.value}
                placeholder={"Please enter the " + props.label.toLowerCase()}
                onChange={props.inputChange}/>
        </div>
    )
}

export default TextField;