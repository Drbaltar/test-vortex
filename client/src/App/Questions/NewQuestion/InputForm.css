.card {
    margin: 20px;
}

.form-group {
    text-align: left;
}

.form-buttons {
    float: right;
}