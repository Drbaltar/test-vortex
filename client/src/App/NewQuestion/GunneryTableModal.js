import React from 'react';
import { Button, Modal, ModalHeader, ModalBody, ModalFooter } from 'reactstrap';
import SelectBox from './Components/SelectBox';
import GunnerySelector from './Components/GunnerySelector';

class GunneryTableModal extends React.Component {
    constructor(props) {
        super(props);

        this.initialState = {
            modal: false,
            unitType: '',
            testType: '',
            table: '',
            subtask: ''
        }

        this.state = this.initialState;

        this.toggle = this.toggle.bind(this);
    };

    toggleTestType = () => {
        let testTypeOptions;
        if (this.state.unitType === 'Battalion') {
            testTypeOptions = ['', 'Tactics', 'Communications', 'RSOP', 'Early Warning/Mission Command'];
        } else if (this.state.unitType === 'Battery') {
            testTypeOptions = ['', 'Tactics', 'Communications', 'RSOP', 'Early Warning/Mission Command', 'Launcher'];
        } else return;

        return(
            <SelectBox label="Test Type" id="testType"
                options={testTypeOptions}
                value={this.state.testType}
                inputChange={(event) => this.handleInputChange(event)}/>
        )
    }

    toggleGunnerySelector = () => {
        if (this.state.testType !== '' && this.state.unitType !== '') {
            return(
                <div>
                    <hr className="mt-3"></hr>
                    <GunnerySelector unitType={this.state.unitType}
                        testType={this.state.testType}
                        subtaskSelect={(table, subtask) => this.handleGunneryChange(table, subtask)}/>
                </div>
            )
        }
    }

    handleInputChange = (event) => {
        const {target: { id, value}} = event;
        this.setState({[id]: value}, () => console.log(this.state));        
    };

    handleGunneryChange = (table, subtask) => {
        this.setState({table: table, subtask: subtask}, () => console.log(this.state));
    };

    toggle = () => {
        if (this.state.modal) {
            this.setState(this.initialState);
        } else {
            this.setState(prevState => ({
                modal: !prevState.modal
        }));
        }
    };

    submitGunneryInfo = () => {
        if (this.state.unitType && this.state.testType && this.state.table && this.state.subtask) {
            let newEntry = {unitType: this.state.unitType,
                testType: this.state.testType,
                table: this.state.table,
                subtask: this.state.subtask}
            this.props.updateGunneryList(newEntry);
            this.toggle();
        }
    };

    render() {
        const testType = this.toggleTestType();
        const gunneryTable = this.toggleGunnerySelector();

        return (
        <div>
            <Button color='info' onClick={this.toggle} className='mt-3 mb-3'>{this.props.buttonLabel}</Button>
            <Modal size="lg" isOpen={this.state.modal} toggle={this.toggle}>
            <ModalHeader toggle={this.toggle}>Select Gunnery Information</ModalHeader>
            <ModalBody>
                <SelectBox label="Unit Type" id="unitType"
                    options={['', 'Battery', 'Battalion']}
                    value={this.state.unitType}
                    inputChange={(event) => this.handleInputChange(event)}/>
                {testType}
                {gunneryTable}
                </ModalBody>
            <ModalFooter>
                <Button color="success" onClick={this.submitGunneryInfo}>Submit</Button>
                <Button color="secondary" onClick={this.toggle}>Cancel</Button>
            </ModalFooter>
            </Modal>
        </div>
        );
  }
}

export default GunneryTableModal;