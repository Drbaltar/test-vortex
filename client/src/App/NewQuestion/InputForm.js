import React from 'react';
import Axios from 'axios';
import TextField from './Components/TextField';
import TextArea from './Components/TextArea';
import SelectBox from './Components/SelectBox';
import FormButtons from './Components/FormButtons';
import GunneryTableModal from './GunneryTableModal';
import './InputForm.css';

class InputForm extends React.Component {
    constructor(props) {
        super(props);

        this.initialState = {
            questionType: 'Multiple Choice',
            questionDescription: '',
            correctAnswer: '',
            answerA: '',
            answerB: '',
            answerC: '',
            testType: '',
            table: '',
            subtask: '',
            topic: ''
        }

        this.state = this.initialState;
    };

    handleInputChange = (event) => {
        const {target: { id, value}} = event;
        this.setState({[id]: value});

        // If the question type is changed, the correct answer field will be cleared
        if (id === 'questionType') {
            this.setState({correctAnswer: ''});
        }
    };

    handleClickEvent = (event) => {
        event.preventDefault();
        if (event.target.id === 'clearAllButton') {
            this.setState(this.initialState);
        } else if (event.target.id === 'submitButton') {
            this.setState({gunneryTable: {table: this.state.table, subtask: this.state.subtask}}, () => {
                Axios.post('http://localhost:5000/api/questions/submit', this.state)
                    .then((response) => console.log(response.status));
            })
        }
    };

    // Return the appropriate correct answer field based on the type of question
    getCorrectAnswerField = () => {
        if (this.state.questionType === 'True or False') {
            return (
                <SelectBox label="Correct Answer"id="correctAnswer"
                    options={['', 'True', 'False']}
                    value={this.state.correctAnswer}
                    inputChange={(event) => this.handleInputChange(event)}/>
            )
        } else {
            return (
                <TextField label="Correct Answer"id="correctAnswer" type="text"
                    value={this.state.correctAnswer}
                    inputChange={(event) => this.handleInputChange(event)}/>
            )
        }
    };

    checkMultipleChoice = () => {
        if (this.state.questionType === 'Multiple Choice') {
            return(
                <div className="border border-dark rounded p-3">
                    <small id="multChoiceAnswerLabel" className="form-text">Input the incorrect answer options below:</small>
                    <TextField label="Answer A"id="answerA" type="text"
                        value={this.state.answerA}
                        inputChange={(event) => this.handleInputChange(event)}/>
                    <TextField label="Answer B"id="answerB" type="text"
                        value={this.state.answerB}
                        inputChange={(event) => this.handleInputChange(event)}/>
                    <TextField label="Answer C"id="answerC" type="text"
                        value={this.state.answerC}
                        inputChange={(event) => this.handleInputChange(event)}/>
                </div>
            )
        };
    };

    render() {
        // Format the correct answer field based on the question type
        let correctAnswerField = this.getCorrectAnswerField();

        // Render additional answer fields if question is Multiple Choice
        let multChoiceAnswers = this.checkMultipleChoice();

        return(
            <form className="card bg-light m-10">
                <h1 className="card-header">Add New Question</h1>
                    <div className="p-4">
                        <SelectBox label="Question Type" id="questionType"
                            options={['Multiple Choice', 'Fill-in-the-Blank', 'True or False']}
                            value={this.state.questionType}
                            inputChange={(event) => this.handleInputChange(event)}/>
                        <TextArea label="Question Description" id="questionDescription" type="text" rows="4"
                            value={this.state.questionDescription}
                            inputChange={(event) => this.handleInputChange(event)}/>
                        {correctAnswerField}
                        {multChoiceAnswers}
                        <GunneryTableModal buttonLabel='Add Gunnery Table/Subtask'/>
                        <TextField label="Topic"id="topic" type="text"
                            value={this.state.topic}
                            inputChange={(event) => this.handleInputChange(event)}/>
                        <FormButtons clickHandler={(event) => this.handleClickEvent(event)}/>
                    </div>
            </form> 
        )
    }
}

export default InputForm;