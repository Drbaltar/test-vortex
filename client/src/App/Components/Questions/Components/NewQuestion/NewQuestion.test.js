import React from 'react';
import Axios from 'axios';
import { shallow } from 'enzyme';
import NewQuestion from './NewQuestion';

describe('NewQuestion', () => {
    const wrapper = shallow(<NewQuestion questionType='Patriot'/>);

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('renders the SubmissionBody component', () => {
        expect(wrapper.find('SubmissionBody').exists()).toBe(true);
    });

    describe('when the questionType prop is Patriot', () => {
        beforeAll(() => {
            wrapper.setProps({ questionType: 'Patriot' });
        });

        it('renders the IBCSQuestionForm', () => {
            expect(wrapper.find('PatriotQuestionForm').exists()).toBe(true);
        });

        it('passes the correct title to the question form', () => {
            expect(wrapper.find('PatriotQuestionForm').prop('title')).toBe('Add New Question (Patriot)');
        });

        it('passes the correct submitMapping object to the SubmissionBody component', () => {
            const expectedSubmitMapping = {
                submitButton: {
                    requestFunction: Axios.post,
                    requestURI: '/api/questions/patriot/pending'
                }
            };

            expect(wrapper.find('SubmissionBody').prop('submitMapping')).toEqual(expectedSubmitMapping);
        });
    });

    describe('when the questionType prop is IBCS', () => {
        beforeAll(() => {
            wrapper.setProps({ questionType: 'IBCS' });
        });

        it('renders the IBCSQuestionForm', () => {
            expect(wrapper.find('IBCSQuestionForm').exists()).toBe(true);
        });

        it('passes the correct title to the question form', () => {
            expect(wrapper.find('IBCSQuestionForm').prop('title')).toBe('Add New Question (IBCS)');
        });

        it('passes the correct submitMapping object to the SubmissionBody component', () => {
            const expectedSubmitMapping = {
                submitButton: {
                    requestFunction: Axios.post,
                    requestURI: '/api/questions/ibcs/pending'
                }
            };

            expect(wrapper.find('SubmissionBody').prop('submitMapping')).toEqual(expectedSubmitMapping);
        });
    });

    it('renders the FormButtons component', () => {
        expect(wrapper.find('FormButtons').exists()).toBe(true);
    });

    it('passes the correct props to the FormButtons component', () => {
        expect(wrapper.find('FormButtons').prop('submit')).toEqual(true);
    });
});