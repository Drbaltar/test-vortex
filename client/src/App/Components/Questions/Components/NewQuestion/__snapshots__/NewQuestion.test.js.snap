// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewQuestion renders correctly 1`] = `
<SubmissionBody
  submitMapping={
    Object {
      "submitButton": Object {
        "requestFunction": [Function],
        "requestURI": "/api/questions/patriot/pending",
      },
    }
  }
>
  <PatriotQuestionForm
    title="Add New Question (Patriot)"
  >
    <FormButtons
      submit={true}
    />
  </PatriotQuestionForm>
</SubmissionBody>
`;
